feat: Installation access tokens for Git operations on GH PLUTO-772 (#1931)
codacy/docs: docs/getting-started/which-permissions-does-codacy-need-from-my-account.md

      <td>Read & Write</td>
      <td>Codacy sets the status of commits according to the result of code analysis.</td>
    </tr>
    <tr>

>       <td>Administration</td>

Author: @andrzej-janczak Using installation access tokens on GitHub, I believe we don't need the `Administration` permission anymore. Can you please confirm? 🙏

Reviewer: Yes, this is true, for sure we drop `write` access 
read access is to be checked in task: https://codacy.atlassian.net/browse/PLUTO-765 
but for 99% will be dropped also, so  I for now I would assume we just drop `Administration`

Author: Ok. Let me know when we finish that task. If we keep `Administration` Read permission, I will need to explain how we use that permission.

Author: @andrzej-janczak I kept the Administration permission for now describing the fallback mechanism. Let me know if it's correct.

Reviewer: LGTM

      <td>Read & Write</td>
      <td>Codacy sets the status of commits according to the result of code analysis.</td>
    </tr>
    <tr>
      <td>Contents</td>
      <td>Read-Only</td>

>       <td>Codacy retrieves repository contents to get installation access tokens, which are used to integrate with your repositories and clone them for code analysis.</td>

Reviewer 2: I believe we could also include `code coverage` in the first line (I'll be honest, I don't fully understand how and why the new coverage pipeline needs this permission, but it does)

I'd also suggest adding an emphasized call to action here for people to approve this permission.
> Codacy is planing on deprecating the need for Repository Administration permissions, but to do so resources from Contents permssion are needed. This Permission has strated being requesteted recently so to future-proof your codacy usage, make sure you that you aprove [approve Codacy GitHub App updated Repository Coontents permission](https://docs.github.com/en/apps/using-github-apps/approving-updated-permissions-for-a-github-app) on your GitHub organization.

>       <td>Codacy uses installation access tokens when integrating with your repositories and cloning them for code quality analysis, and retrieves repository contents to provide faster code coverage analysis.</td>

Author: @andrzej-janczak @lorandszakacs I did a tweak here to mention **installation access tokens**. Let me know if it's correct. 🙏 
I want to confirm if Codacy **uses** installation access tokens. (instead of creating/generating... for that maybe we would need more than read-only access, right?)

Reviewer: look ok 👍

>       <td>Codacy uses installation access tokens when integrating with your repositories and cloning them for code quality analysis, and retrieves repository contents to provide faster code coverage analysis.</td>

Reviewer 3: Not sure I'm getting the meaning of this sentence. I'm attempting a reword, let me know if it's correct:

```suggestion
      <td>Codacy uses installation access tokens when integrating with your repositories to clone them for code quality analysis and to retrieve their contents for faster code coverage analysis.</td>
```

Author: > Codacy uses installation access tokens when integrating with your repositories to clone them for code quality analysis and to retrieve their contents for faster code coverage analysis.

@nicklem This way we are saying that the new Coverage engine also uses installation access tokens... Does it use tokens? 🤔 
I based on the [original sentence](https://github.com/codacy/docs/pull/1931/files#diff-47d9d57a1ef035b843d334b8ceabbfee55e41dc6fe3e45012197b57631db1c28L74), which does not specify if Contents permissions is required to use tokens or something else...

Reviewer 3: Oh, I see. I got the meaning wrong. The fact is that the fragment at the beginning, "Codacy uses installation access tokens", seems to imply that the installation access tokens are used to access contents, but it's the other way around, i.e. the Contents permission is used to access the installation access tokens?

If so, maybe rewording like this can work?

Either:
```
 <td>Codacy retrieves repository contents to get installation access tokens, which are used to integrate with your repositories and clone them for code quality analysis, and to provide faster code coverage analysis.</td>
```

Or:
```
<td>Codacy retrieves repository contents to provide faster code coverage analysis and to retrieve installation access tokens, which are used to integrate with your repositories and clone them for code quality analysis</td>
```

Reviewer 3: I wonder if we could simplify further, avoid mentioning "faster code coverage analysis", and simply group that into a generic "clone them for code analysis":

```
<td>Codacy retrieves repository contents to get installation access tokens, which are used to integrate with your repositories and clone them for code analysis.</td>
```

Reviewer 3: I wonder if we could simplify further, avoid mentioning "faster code coverage analysis", and simply group that into a generic "clone them for code analysis":

```suggestion
      <td>Codacy retrieves repository contents to get installation access tokens, which are used to integrate with your repositories and clone them for code analysis.</td>
```

Reviewer 3: Perhaps we can simplify even further:

```suggestion
      <td>Codacy retrieves repository contents to get installation access tokens, which are used to integrate with your repositories and clone them for code analysis.</td>
```

    </tr>
    <tr>
      <td>Administration</td>
      <td>Read & Write</td>

>       <td>This permission will soon be removed and is currently used <strong>exclusively as a fallback mechanism</strong> when the Contents permission is missing. In this case, Codacy <a href="#why-does-codacy-ask-for-permission-to-create-ssh-keys">creates an SSH key on the repository</a> to allow cloning and integrating with your repository.<br/>To avoid disruptions in the future, make sure you accept Codacy GitHub App updated permissions on your GitHub organization.</td>

Reviewer: maybe we could add link to GH docs 
https://docs.github.com/en/apps/using-github-apps/approving-updated-permissions-for-a-github-app

>       <td>This permission will soon be removed and is currently used <strong>exclusively as a fallback mechanism</strong> when the Contents permission is missing. In this case, Codacy <a href="#why-does-codacy-ask-for-permission-to-create-ssh-keys">creates an SSH key on the repository</a> to allow cloning and integrating with your repository.<br/>To avoid disruptions in the future, make sure you [approve Codacy GitHub App updated permissions](https://docs.github.com/en/apps/using-github-apps/approving-updated-permissions-for-a-github-app) on your GitHub organization.</td>

Author: ```suggestion
      <td>This permission will soon be removed and is currently used <strong>exclusively as a fallback mechanism</strong> when the Contents permission is missing. In this case, Codacy <a href="#why-does-codacy-ask-for-permission-to-create-ssh-keys">creates an SSH key on the repository</a> to allow cloning and integrating with your repository.<br/>To avoid disruptions in the future, make sure you <a href="https://docs.github.com/en/apps/using-github-apps/approving-updated-permissions-for-a-github-app">approve Codacy GitHub App updated permissions</a> on your GitHub organization.</td>
```

>       <td>This permission will soon be removed and is currently used <strong>exclusively as a fallback mechanism</strong> when the Contents permission is missing. In this case, Codacy <a href="#why-does-codacy-ask-for-permission-to-create-ssh-keys">creates an SSH key on the repository</a> to allow cloning and integrating with your repository.<br/>To avoid disruptions in the future, make sure you <a href="https://docs.github.com/en/apps/using-github-apps/approving-updated-permissions-for-a-github-app">approve Codacy GitHub App updated permissions</a> on your GitHub organization.</td>

Reviewer 2: I would suggest the emphasis (strong) in this text should be on _the soon to be deprecated permission_, rather than the current use we have for it.
ex: `<strong>This permission will soon be removed</strong> and is currently used exclusively as a fallback mechanism (...)`

Also, as we are still figuring out possible issues and ironing out details as we go, it's gonna be inherently difficult to align a "release date" for this, so I'd suggest we adapt language to accommodate for this, something like avoiding the term "exclusively": 
> This permission will soon be removed and is currently used when it's not possible to leverage the Repository Contents permission. In this case, Codacy creates an SSH key on the repository to allow cloning and integrating with your repository.
> To future-proof your codacy usage, make sure you that you aprove [approve Codacy GitHub App updated Repository Coontents permission](https://docs.github.com/en/apps/using-github-apps/approving-updated-permissions-for-a-github-app) on your GitHub organization.

It's possible that I'm being creative with the usage of "not possible", I was trying, without lying to encapsulate both the cases that we reverted the behaviour for some testing reason as well as the cases where customers have not accepted the Repository `Contents` permission.

(Also please don't use any of my suggestions verbatim, I'm sure there are better ways to write these things.)

Author: > It's possible that I'm being creative with the usage of "not possible", I was trying, without lying to encapsulate both the cases that we reverted the behaviour for some testing reason as well as the cases where customers have not accepted the Repository Contents permission.

I rephrased it according to the current state of the art. 😉

## Why does Codacy ask for permission to create SSH keys?

!!! info "This section applies only to GitLab and Bitbucket"

On GitLab and Bitbucket organizations, when you add a private repository to Codacy, Codacy uses the integration with your Git provider to create a new SSH key on the repository. Codacy then uses that SSH key every time it needs to clone the repository.

**Codacy only adds read-only SSH keys** and can't access any of your existing SSH keys. You have full control over which organizations and repositories Codacy is authorized to access. Codacy doesn't change the contents or member privileges of any repository you authorize it to analyze.

We understand the desire for security and privacy and find that the SSH protocol is preferable to HTTPS as it separates Codacy's access rights from the one of the users.

> !!! tip

Author: @andrzej-janczak Can you please validate if these are the most indicated links to GitLab and Bitbucket documentation?

Reviewer: hm for GL we use deploy keys that are located here
doc: https://docs.gitlab.com/ee/user/project/deploy_keys/index.html#view-deploy-keys

![Screenshot 2023-11-14 at 09 52 54](https://github.com/codacy/docs/assets/122018786/6e731aee-5c5c-4993-ae86-7f17e20da2ca)

for BB looks like the best place would be: 
https://support.atlassian.com/bitbucket-cloud/docs/configure-repository-settings/ 
![Screenshot 2023-11-14 at 09 55 23](https://github.com/codacy/docs/assets/122018786/92a4e05e-a294-49d1-8f7d-fc9f513986e9)

this is also like a deploy key (read only ssh key per repo) but not using "deploy" word.

Author: Updated!

## Why does Codacy ask for permission to create SSH keys?

!!! info "This section applies only to GitLab and Bitbucket"

>     For GitHub, Codacy creates an SSH key on the repository **exclusively as a fallback mechanism** when the Contents permission is missing. To avoid disruptions in the future, make sure you accept Codacy GitHub App [updated permissions](#github-cloud) on your GitHub organization.

Reviewer 3: Does this admonition break our usual structure? I remember using custom titles only with empty `info` admonitions.

Either way, I'm OK with using it like this here. This is temporary I presume?

Author: You are right, this is the first time we add additional content to an `info` admonition. But at the same time, we always use `info` admonitions to say "This page/section only applies to <something>"...

So I opted to break our usual standard by adding more information (because it's temporary), instead of breaking by using a different admonition type. 🤓 

I would keep it like this, as we'll be removing the additional temporary content in short/medium future.

WDYT?

Reviewer 3: Sounds good, especially if this is temporary 👍

>     For GitHub, Codacy creates an SSH key on the repository **exclusively as a fallback mechanism** when the Contents permission is missing. To avoid disruptions in the future, make sure you accept Codacy GitHub App [updated permissions](#github-cloud) on your GitHub organization.

Reviewer 2: > For Github, Codacy only uses an SSH key when it's not possible to leverage the Repository Contents permission. In this case, Codacy uses a repository SSH key allow cloning and integrating with your repository. As we progress with our plan to deprecate Repository Administration permissions we will fully stop to create, use and support SSH keys. To avoid disruptions in the future, make sure you accept Codacy GitHub App [updated permissions](#github-cloud) on your GitHub organization.

`Codacy creates an SSH key on the repository` this sentence is not true yet. SSH keys are created when a repository is added, and we will only stop this behaviour when we fully deprecate the `Administration` permission.
